File: python-sdk/src/mcpbuilder/websocket_client.py
"""WebSocket client for connecting to MCP client service."""

import asyncio
import json
import logging
from typing import Callable, Optional, List
from urllib.parse import urlencode
import websockets
from websockets.client import WebSocketClientProtocol

from .types import ServerEvent, ClientMessage, HistoryMessage

logger = logging.getLogger(__name__)


class WebSocketClient:
    """WebSocket client for MCP chat service communication."""
    
    def __init__(self):
        """Initialize the WebSocket client."""
        self.ws: Optional[WebSocketClientProtocol] = None
        self.is_connected = False
        self.was_ever_connected = False
        self.message_callbacks: list[Callable[[ServerEvent], None]] = []
        self.reconnect_attempts = 0
        self.max_reconnect_attempts = 3
        self.auto_reconnect_enabled = False
        self.reconnect_delay_ms = 2000
        self.url = ""
        self.connection_params = {}
        self._reconnect_task: Optional[asyncio.Task] = None
        self._receive_task: Optional[asyncio.Task] = None
        self._initialized_event: Optional[asyncio.Event] = None

    async def connect(
        self,
        url: str,
        deployment_name: str,
        project_token: str,
        auto_reconnect: bool = True,
        max_reconnect_attempts: int = 3,
        reconnect_delay_ms: int = 2000,
        timezone: Optional[str] = None,
        system_message: Optional[str] = None,
        language: Optional[str] = None,
        security_params: Optional[dict] = None,
        message_history: Optional[List[HistoryMessage]] = None,
        cache_history: bool = False,
    ) -> None:
        """
        Connect to the MCP WebSocket service.
        
        Args:
            url: Base WebSocket URL (e.g., ws://localhost:8000)
            deployment_name: Deployment name for authentication
            project_token: Project authentication token
            auto_reconnect: Enable automatic reconnection
            max_reconnect_attempts: Maximum reconnection attempts
            reconnect_delay_ms: Base delay for reconnection
            timezone: Optional timezone for the session
            system_message: Optional system message/prompt for the LLM
            language: Optional language setting for the assistant
            security_params: Optional dict of security parameters to merge with
                loaded security headers (these take precedence)
            message_history: Optional initial message history to provide context
            cache_history: Whether to cache conversation history on the server (default: False)
        """
        if not deployment_name:
            raise ValueError("deployment_name is required")
        if not project_token:
            raise ValueError("project_token is required")
        
        self.url = url
        self.auto_reconnect_enabled = auto_reconnect
        self.max_reconnect_attempts = max_reconnect_attempts
        self.reconnect_delay_ms = reconnect_delay_ms
        self.connection_params = {
            "deployment_name": deployment_name,
            "project_token": project_token,
        }

        if timezone:
            self.connection_params["timezone"] = timezone
        if system_message:
            self.connection_params["system_message"] = system_message
        if language:
            self.connection_params["language"] = language
        if security_params:
            # JSON-encode security params for URL transport
            # Filter out None and empty values to prevent server-side errors
            valid_params = {k: v for k, v in security_params.items() if v is not None and v != ""}
            if valid_params:
                self.connection_params["security_params"] = json.dumps(valid_params)
        if message_history:
            # JSON-encode message history for URL transport
            history_dicts = [msg.to_dict() for msg in message_history]
            self.connection_params["message_history"] = json.dumps(history_dicts)
        # Always include cache_history parameter
        self.connection_params["cache_history"] = str(cache_history).lower()
        
        await self._connect_internal()

    async def _connect_internal(self) -> None:
        """Internal connection logic."""
        try:
            ws_url = f"{self.url}/ws/chat?{urlencode(self.connection_params)}"
            logger.info(f"Connecting to {ws_url}")
            
            # Create an event to wait for the "initialized" status from server
            self._initialized_event = asyncio.Event()
            
            self.ws = await websockets.connect(ws_url)
            self.is_connected = True
            self.was_ever_connected = True
            self.reconnect_attempts = 0
            
            logger.info("Connected to MCP service")
            
            # Start message receiving loop as a background task (don't await it)
            self._receive_task = asyncio.create_task(self._receive_loop())
            
            # Wait for the server to send "initialized" status (with timeout)
            try:
                await asyncio.wait_for(self._initialized_event.wait(), timeout=30.0)
                logger.info("Session initialized by server")
            except asyncio.TimeoutError:
                logger.warning("Timeout waiting for server initialization, proceeding anyway")
            
        except Exception as e:
            logger.error(f"Connection error: {e}")
            self.is_connected = False
            
            if self.ws:
                await self.ws.close()
                self.ws = None
            
            # Try to reconnect
            if self.was_ever_connected and self.auto_reconnect_enabled:
                await self._handle_disconnect()
            else:
                self._broadcast_event(ServerEvent(
                    type="error",
                    error=f"Connection failed: {str(e)}"
                ))

    async def _receive_loop(self) -> None:
        """Receive messages from the server."""
        if not self.ws:
            return
        
        try:
            async for message in self.ws:
                try:
                    data = json.loads(message)
                    
                    # Handle service-initiated closing
                    if data.get("type") == "closing":
                        logger.info(f"Service closing: {data.get('reason')}")
                        self._broadcast_event(ServerEvent(
                            type="closing",
                            reason=data.get("reason")
                        ))
                        break
                    
                    # Check for "initialized" status to signal connection is ready
                    if data.get("type") == "status" and data.get("message") == "initialized":
                        logger.info("Received initialized status from server")
                        if self._initialized_event:
                            self._initialized_event.set()
                    
                    # Convert to ServerEvent
                    event = ServerEvent.from_dict(data)
                    self._broadcast_event(event)
                    
                except json.JSONDecodeError as e:
                    logger.error(f"Failed to parse message: {e}")
                except Exception as e:
                    logger.error(f"Error processing message: {e}")
        
        except asyncio.CancelledError:
            logger.debug("Receive loop cancelled")
            raise
        except Exception as e:
            logger.error(f"Receive loop error: {e}")
        finally:
            self.is_connected = False
            await self._handle_disconnect()

    async def _handle_disconnect(self) -> None:
        """Handle disconnection and attempt reconnection."""
        logger.info("Handling disconnect")
        self._broadcast_event(ServerEvent(
            type="status",
            status="Disconnected from chat service"
        ))
        
        if (
            self.was_ever_connected
            and self.auto_reconnect_enabled
            and self.reconnect_attempts < self.max_reconnect_attempts
        ):
            self.reconnect_attempts += 1
            delay = (self.reconnect_delay_ms * self.reconnect_attempts) / 1000.0
            logger.info(
                f"Reconnecting in {delay}s "
                f"({self.reconnect_attempts}/{self.max_reconnect_attempts})"
            )
            
            await asyncio.sleep(delay)
            await self._connect_internal()

    async def send_message(
        self,
        text: str,
        message_history: Optional[List[HistoryMessage]] = None,
    ) -> None:
        """
        Send a chat message to the service.
        
        Args:
            text: Message text to send
            message_history: Optional message history override for this message only
        """
        if not self.is_connected or not self.ws:
            logger.error("WebSocket not connected")
            self._broadcast_event(ServerEvent(
                type="error",
                error="Not connected to chat service"
            ))
            return
        
        try:
            history_dicts = None
            if message_history is not None:
                history_dicts = [msg.to_dict() for msg in message_history]

            message = ClientMessage(
                type="message",
                text=text,
                message_history=history_dicts,
            )
            await self.ws.send(json.dumps(message.to_dict()))
            logger.debug(f"Sent message: {text[:50]}...")
        except Exception as e:
            logger.error(f"Failed to send message: {e}")
            self._broadcast_event(ServerEvent(
                type="error",
                error=f"Failed to send message: {str(e)}"
            ))

    async def send_stop(self) -> None:
        """Send a stop request to cancel the current operation."""
        if not self.is_connected or not self.ws:
            logger.error("WebSocket not connected")
            return
        
        try:
            message = ClientMessage(type="stop")
            await self.ws.send(json.dumps(message.to_dict()))
            logger.info("Sent stop request")
        except Exception as e:
            logger.error(f"Failed to send stop: {e}")

    async def send_tool_consent(
        self,
        consent_id: str,
        allow: bool,
        allow_all: bool = False,
    ) -> None:
        """
        Respond to a tool consent request.
        
        Args:
            consent_id: Consent request ID
            allow: Whether to allow the tool execution
            allow_all: Whether to always allow this tool in the future
        """
        if not self.is_connected or not self.ws:
            logger.error("WebSocket not connected")
            return
        
        try:
            message = ClientMessage(
                type="tool_consent_response",
                consent_id=consent_id,
                allow=allow,
                allow_all=allow_all if allow_all else None,
            )
            await self.ws.send(json.dumps(message.to_dict()))
            logger.info(
                f"Sent tool consent: {allow} "
                f"(allow_all={allow_all}) for {consent_id}"
            )
        except Exception as e:
            logger.error(f"Failed to send tool consent: {e}")

    async def send_consent_set(
        self,
        allow_all: bool = False,
        tool_name: Optional[str] = None,
    ) -> None:
        """
        Set consent preferences.
        
        Args:
            allow_all: If True, auto-allow all tool calls
            tool_name: If provided, auto-allow this specific tool
        """
        if not self.is_connected or not self.ws:
            logger.error("WebSocket not connected")
            return
        
        try:
            message = {"type": "tool_consent_set"}
            if allow_all:
                message["allow_all"] = True
            elif tool_name:
                message["tool_name"] = tool_name
            
            await self.ws.send(json.dumps(message))
            logger.info(f"Sent consent set: allow_all={allow_all}, tool_name={tool_name}")
        except Exception as e:
            logger.error(f"Failed to send consent set: {e}")

    def on_message(self, callback: Callable[[ServerEvent], None]) -> None:
        """
        Register a callback for server events.
        
        Args:
            callback: Function to call when events are received
        """
        self.message_callbacks.append(callback)

    def clear_callbacks(self) -> None:
        """Clear all registered callbacks."""
        self.message_callbacks.clear()

    def _broadcast_event(self, event: ServerEvent) -> None:
        """Broadcast event to all registered callbacks."""
        for callback in self.message_callbacks:
            try:
                callback(event)
            except Exception as e:
                logger.error(f"Error in callback: {e}")

    async def disconnect(self) -> None:
        """Disconnect from the service."""
        self.is_connected = False
        self.was_ever_connected = False
        self.reconnect_attempts = 0
        self.clear_callbacks()
        
        # Cancel the receive task if running
        if self._receive_task and not self._receive_task.done():
            self._receive_task.cancel()
            try:
                await self._receive_task
            except asyncio.CancelledError:
                pass
            self._receive_task = None
        
        if self.ws:
            try:
                await self.ws.close()
            except Exception as e:
                logger.error(f"Error closing WebSocket: {e}")
            finally:
                self.ws = None
        
        logger.info("WebSocket disconnected")

    def is_connected_status(self) -> bool:
        """Check if currently connected."""
        return self.is_connected and self.ws is not None

File: typescript-sdk/src/websocket-client.ts
/**
 * WebSocket client for connecting to MCP client service.
 * 
 * This module is isomorphic - it works in both Node.js and browser environments.
 * - In Node.js: uses the 'ws' package
 * - In browser: uses native WebSocket
 */

import type {
  ServerEvent,
  WebSocketConnectionParams,
  HistoryMessage,
} from './types.js';
import {
  parseServerEvent,
  createMessage,
  createStopMessage,
  createConsentResponse,
  createConsentSet,
} from './types.js';

// Declare global types for browser environment detection
declare const globalThis: {
  WebSocket?: new (url: string) => WebSocket;
} & typeof global;

// Detect if we're in a browser environment
const isBrowser = typeof globalThis.WebSocket !== 'undefined';

// WebSocket type that works in both environments
type WebSocketLike = {
  readyState: number;
  onopen: ((event: unknown) => void) | null;
  onmessage: ((event: { data: unknown }) => void) | null;
  onerror: ((event: unknown) => void) | null;
  onclose: ((event: unknown) => void) | null;
  send(data: string): void;
  close(): void;
};

/**
 * Get the WebSocket constructor for the current environment.
 */
async function getWebSocketConstructor(): Promise<new (url: string) => WebSocketLike> {
  if (isBrowser) {
    return globalThis.WebSocket as unknown as new (url: string) => WebSocketLike;
  } else {
    // Dynamic import for Node.js
    const ws = await import('ws');
    return ws.default as unknown as new (url: string) => WebSocketLike;
  }
}

/**
 * Logger interface for the WebSocket client.
 */
interface Logger {
  debug(message: string): void;
  info(message: string): void;
  warn(message: string): void;
  error(message: string): void;
}

/**
 * Default console logger.
 */
const defaultLogger: Logger = {
  debug: (msg) => console.debug(`[MCPBuilder] ${msg}`),
  info: (msg) => console.info(`[MCPBuilder] ${msg}`),
  warn: (msg) => console.warn(`[MCPBuilder] ${msg}`),
  error: (msg) => console.error(`[MCPBuilder] ${msg}`),
};

/**
 * WebSocket client for MCP chat service communication.
 */
export class WebSocketClient {
  private ws: WebSocketLike | null = null;
  private _isConnected = false;
  private _isConnecting = false;
  private wasEverConnected = false;
  private messageCallbacks: Array<(event: ServerEvent) => void> = [];
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 3;
  private autoReconnectEnabled = false;
  private reconnectDelayMs = 2000;
  private url = '';
  private connectionParams: Record<string, string> = {};
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private initializedResolve: (() => void) | null = null;
  private logger: Logger;

  constructor(logger?: Logger) {
    this.logger = logger || defaultLogger;
  }

  /**
   * Check if currently connected.
   * Note: WebSocket.OPEN = 1 in both browser and Node.js
   */
  get isConnected(): boolean {
    return this._isConnected && this.ws !== null && this.ws.readyState === 1;
  }

  /**
   * Connect to the MCP WebSocket service.
   */
  async connect(params: WebSocketConnectionParams): Promise<void> {
    const {
      url,
      deploymentName,
      projectToken,
      autoReconnect = true,
      maxReconnectAttempts = 3,
      reconnectDelayMs = 2000,
      timezone,
      systemMessage,
      language,
      securityParams,
      messageHistory,
      cacheHistory,
    } = params;

    if (!deploymentName) {
      throw new Error('deploymentName is required');
    }
    if (!projectToken) {
      throw new Error('projectToken is required');
    }

    this.url = url;
    this.autoReconnectEnabled = autoReconnect;
    this.maxReconnectAttempts = maxReconnectAttempts;
    this.reconnectDelayMs = reconnectDelayMs;
    this.connectionParams = {
      deployment_name: deploymentName,
      project_token: projectToken,
    };

    if (timezone) {
      this.connectionParams.timezone = timezone;
    }
    if (systemMessage) {
      this.connectionParams.system_message = systemMessage;
    }
    if (language) {
      this.connectionParams.language = language;
    }
    if (securityParams && Object.keys(securityParams).length > 0) {
      // JSON-encode security params for URL transport (matching Python SDK)
      // Filter out null, undefined, and empty string values to prevent server-side errors
      const validParams: Record<string, string> = {};
      for (const [key, value] of Object.entries(securityParams)) {
        if (value !== null && value !== undefined && value !== '') {
          validParams[key] = value;
        }
      }
      if (Object.keys(validParams).length > 0) {
        this.connectionParams.security_params = JSON.stringify(validParams);
      }
    }
    if (messageHistory && messageHistory.length > 0) {
      // JSON-encode message history for URL transport
      this.connectionParams.message_history = JSON.stringify(messageHistory);
    }
    if (cacheHistory !== undefined) {
      this.connectionParams.cache_history = String(cacheHistory);
    }

    await this.connectInternal();
  }

  /**
   * Internal connection logic.
   */
  private async connectInternal(): Promise<void> {
    // Guard against concurrent connections
    if (this._isConnecting) {
      this.logger.warn('Connection already in progress, skipping');
      return;
    }
    if (this._isConnected) {
      this.logger.warn('Already connected, skipping');
      return;
    }
    
    this._isConnecting = true;
    
    const queryParams = new URLSearchParams(this.connectionParams);
    const wsUrl = `${this.url}/ws/chat?${queryParams.toString()}`;
    this.logger.info(`Connecting to ${wsUrl}`);

    // Get the appropriate WebSocket constructor for the environment
    const WebSocketImpl = await getWebSocketConstructor();

    return new Promise((resolve, reject) => {
      try {
        // Set up initialization resolver
        this.initializedResolve = null;
        new Promise<void>((res) => {
          this.initializedResolve = res;
        });

        this.ws = new WebSocketImpl(wsUrl);

        this.ws.onopen = () => {
          this._isConnected = true;
          this.wasEverConnected = true;
          this.reconnectAttempts = 0;
          this.logger.info('Connected to MCP service');
        };

        this.ws.onmessage = (event: { data: unknown }) => {
          try {
            // Handle both browser and Node.js message data
            const rawData = typeof event.data === 'string' ? event.data : String(event.data);
            const data = JSON.parse(rawData) as Record<string, unknown>;

            // Handle service-initiated closing
            if (data.type === 'closing') {
              this.logger.info(`Service closing: ${data.reason}`);
              this.broadcastEvent({
                type: 'closing',
                reason: data.reason as string,
              });
              return;
            }

            // Check for "initialized" status to signal connection is ready
            if (data.type === 'status' && data.message === 'initialized') {
              this.logger.info('Received initialized status from server');
              this._isConnecting = false; // Connection complete
              if (this.initializedResolve) {
                this.initializedResolve();
                this.initializedResolve = null; // Clear to prevent timeout from re-resolving
              }
              resolve();
            }

            // Convert to ServerEvent
            const serverEvent = parseServerEvent(data);
            if (serverEvent.type === 'final') {
              this.logger.info(`Received 'final' event from server`);
            }
            this.broadcastEvent(serverEvent);
          } catch (err) {
            this.logger.error(`Failed to parse message: ${err}`);
          }
        };

        this.ws.onerror = (error: unknown) => {
          const errorMessage = error && typeof error === 'object' && 'message' in error
            ? (error as { message?: string }).message || 'Unknown error'
            : 'Unknown error';
          this.logger.error(`WebSocket error: ${errorMessage}`);
          if (!this.wasEverConnected) {
            reject(new Error(`Connection failed: ${errorMessage}`));
          }
        };

        this.ws.onclose = () => {
          this._isConnected = false;
          this.handleDisconnect();
        };

        // Set timeout for initialization
        setTimeout(() => {
          if (!this._isConnected) {
            this._isConnecting = false;
            reject(new Error('Connection timeout'));
          } else if (this.initializedResolve) {
            // Resolve anyway if connected but no init message
            this.logger.warn('Timeout waiting for server initialization, proceeding anyway');
            this._isConnecting = false;
            this.initializedResolve();
            this.initializedResolve = null; // Clear to prevent double-resolve
            resolve();
          }
        }, 30000);
      } catch (err) {
        this.logger.error(`Connection error: ${err}`);
        this._isConnected = false;
        this._isConnecting = false;

        if (this.ws) {
          this.ws.close();
          this.ws = null;
        }

        if (this.wasEverConnected && this.autoReconnectEnabled) {
          this.handleDisconnect();
          resolve();
        } else {
          this.broadcastEvent({
            type: 'error',
            error: `Connection failed: ${err}`,
          });
          reject(err);
        }
      }
    });
  }

  /**
   * Handle disconnection and attempt reconnection.
   */
  private handleDisconnect(): void {
    this.logger.info('Handling disconnect');
    this.broadcastEvent({
      type: 'status',
      status: 'Disconnected from chat service',
    });

    if (
      this.wasEverConnected &&
      this.autoReconnectEnabled &&
      this.reconnectAttempts < this.maxReconnectAttempts
    ) {
      this.reconnectAttempts++;
      const delay = (this.reconnectDelayMs * this.reconnectAttempts);
      this.logger.info(
        `Reconnecting in ${delay}ms (${this.reconnectAttempts}/${this.maxReconnectAttempts})`
      );

      this.reconnectTimer = setTimeout(() => {
        this.connectInternal().catch((err) => {
          this.logger.error(`Reconnection failed: ${err}`);
        });
      }, delay);
    }
  }

  /**
   * Send a chat message to the service.
   */
  async sendMessage(text: string, messageHistory?: HistoryMessage[]): Promise<void> {
    if (!this.isConnected || !this.ws) {
      this.logger.error('WebSocket not connected');
      this.broadcastEvent({
        type: 'error',
        error: 'Not connected to chat service',
      });
      return;
    }

    try {
      const message = createMessage(text, messageHistory);
      this.ws.send(JSON.stringify(message));
      this.logger.debug(`Sent message: ${text.substring(0, 50)}...`);
    } catch (err) {
      this.logger.error(`Failed to send message: ${err}`);
      this.broadcastEvent({
        type: 'error',
        error: `Failed to send message: ${err}`,
      });
    }
  }

  /**
   * Send a stop request to cancel the current operation.
   */
  async sendStop(): Promise<void> {
    if (!this.isConnected || !this.ws) {
      this.logger.error('WebSocket not connected');
      return;
    }

    try {
      const message = createStopMessage();
      this.ws.send(JSON.stringify(message));
      this.logger.info('Sent stop request');
    } catch (err) {
      this.logger.error(`Failed to send stop: ${err}`);
    }
  }

  /**
   * Respond to a tool consent request.
   */
  async sendToolConsent(
    consentId: string,
    allow: boolean,
    allowAll = false
  ): Promise<void> {
    if (!this.isConnected || !this.ws) {
      this.logger.error('WebSocket not connected');
      return;
    }

    try {
      const message = createConsentResponse(
        consentId,
        allow,
        allowAll ? allowAll : undefined
      );
      this.ws.send(JSON.stringify(message));
      this.logger.info(
        `Sent tool consent: ${allow} (allow_all=${allowAll}) for ${consentId}`
      );
    } catch (err) {
      this.logger.error(`Failed to send tool consent: ${err}`);
    }
  }

  /**
   * Set consent preferences.
   */
  async sendConsentSet(options: {
    allowAll?: boolean;
    toolName?: string;
  }): Promise<void> {
    if (!this.isConnected || !this.ws) {
      this.logger.error('WebSocket not connected');
      return;
    }

    try {
      const message = createConsentSet(options);
      this.ws.send(JSON.stringify(message));
      this.logger.info(
        `Sent consent set: allow_all=${options.allowAll}, tool_name=${options.toolName}`
      );
    } catch (err) {
      this.logger.error(`Failed to send consent set: ${err}`);
    }
  }

  /**
   * Register a callback for server events.
   */
  onMessage(callback: (event: ServerEvent) => void): void {
    this.messageCallbacks.push(callback);
  }

  /**
   * Clear all registered callbacks.
   */
  clearCallbacks(): void {
    this.messageCallbacks = [];
  }

  /**
   * Broadcast event to all registered callbacks.
   */
  private broadcastEvent(event: ServerEvent): void {
    if (event.type === 'final') {
      this.logger.info(`Broadcasting final event to ${this.messageCallbacks.length} callbacks`);
    }
    for (const callback of this.messageCallbacks) {
      try {
        callback(event);
      } catch (err) {
        this.logger.error(`Error in callback: ${err}`);
      }
    }
  }

  /**
   * Disconnect from the service.
   */
  async disconnect(): Promise<void> {
    this._isConnected = false;
    this._isConnecting = false;
    this.wasEverConnected = false;
    this.reconnectAttempts = 0;
    this.initializedResolve = null;
    this.clearCallbacks();

    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    if (this.ws) {
      try {
        this.ws.close();
      } catch (err) {
        this.logger.error(`Error closing WebSocket: ${err}`);
      } finally {
        this.ws = null;
      }
    }

    this.logger.info('WebSocket disconnected');
  }
}
